### PR TITLE
MESH-1170/ POLYX vesting smart contract

### PR DESCRIPTION
**Pull request by JMoore96**
### modified API

- Created vesting smart contract

### CODE REVIEW

File: contracts/polyx-vesting/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[package]
+name = "vesting"
```
Comment (Neopallium) [inline contracts/polyx-vesting/Cargo.toml]: Update the name here and below to `polyx-vesting`.

File: contracts/polyx-vesting/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+version = "0.1.0"
+authors = ["[your_name] <[your_email]>"]
```
Comment (Neopallium) [inline contracts/polyx-vesting/Cargo.toml]: You can use your name and email address here or:
```
authors = ["Polymesh Association"]
```

File: contracts/polyx-vesting/lib.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    /// We test a simple use case of our contract.
+    #[ink::test]
+    fn it_works() {
+        let mut vesting = Vesting::new(false);
+        assert_eq!(vesting.get(), false);
+        vesting.flip();
+        assert_eq!(vesting.get(), true);
+    }
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: Please update this test code for testing this contract.

**Review by Neopallium — CHANGES_REQUESTED** [review]
Please add a README.md file to describe how to use the contract.

File: contracts/polyx-vesting/lib.rs (hunk 3 above)
Comment (JMoore96) [reply contracts/polyx-vesting/lib.rs]: Alright, working to finish that today.

File: contracts/polyx-vesting/lib.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        #[ink(message)]
+        pub fn get_balance(&self) -> Balance {
+            self.env().balance()
+        }
+
+        /// Release the native token (POLYX) that have already vested.
+        #[ink(message)]
+        pub fn release(&mut self) -> Result<()> {
+            let amount = self.releasable();
+            assert!(amount > 0, "insufficient funds!");
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: Should return an error here to tell the caller that funds are not releasable yet.

File: contracts/polyx-vesting/lib.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            assert_eq!(polyx_vesting.released(), 0u128);
+        }
+
+        #[ink::test]
+        fn vesting_works() {
+            // Constructor works.
+            let mut polyx_vesting = PolyxVesting::new(AccountId::from([0x01; 32]), 1, 3);
+
+            // Release Polyx
+            assert_eq!(polyx_vesting.release(), Ok(()));
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs:214]: The vesting schedule uses the block timestamp, to calculate how much is releasable.  Right here the timestamp will be zero, since no blocks have been "produced", so the releasable amount is zero.  Need to advance to the next block to release some POLYX in the vesting schedule.

There are Ink! testing methods that can be used to get the contract address, advance to the next block, set account balances, etc...: https://docs.rs/ink_env/latest/ink_env/test/index.html

I good example of Ink! contract testing can be seen here: https://github.com/paritytech/ink/blob/v3.4.0/examples/erc20/lib.rs

Please expand this test to check:

1. Calling `release()` before the start of the vesting period returns an error (right now it panics).
2. Calling `release()` after advancing past the start will release funds.  Also check the beneficiary account balance before and after to make sure they receive the correct funds.
3. Calling `release()` from a non-beneficiary account, still only releases funds to the beneficiary and not the caller.
4. Advance the block timestamp into the next vesting period to release more funds and call `release()` again.

File: contracts/polyx-vesting/lib.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+        /// Imports `ink_lang` so we can use `#[ink::test]`.
+        use ink_lang as ink;
+
+        /// We test if the new constructor does its job.
+        #[ink::test]
+        fn new_works() {
+            // Constructor works.
+            let polyx_vesting = PolyxVesting::new(AccountId::from([0x01; 32]), 5, 20);
+            // Ensure the values are stored correctly
+            assert_eq!(polyx_vesting.beneficiary(), AccountId::from([0x01; 32]));
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: Please use the default accounts (Alice, Bob, Charlie, etc...) instead of raw bytes.  You can get them from: https://docs.rs/ink_env/latest/ink_env/test/fn.default_accounts.html

**Review by Neopallium — CHANGES_REQUESTED** [review]

File: contracts/polyx-vesting/lib.rs (hunk 5 above)
Comment (Neopallium) [reply contracts/polyx-vesting/lib.rs:214]: Also note that the `block_timestamp` during testing start at zero and increases by `6` each time the chain is advanced (`ink_env::test::advance_block`).

To test the vesting schedule it would be better to use larger `start` and `duration` numbers.

Comment (JMoore96) [reply contracts/polyx-vesting/lib.rs:214]: working on these changes now.

**Review by Neopallium — CHANGES_REQUESTED** [review]
Please add a README.md file describing how to deploy the contract and use it.

File: contracts/polyx-vesting/lib.rs [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        #[ink(message)]
+        pub fn get_time(&self) -> Balance {
+            self.env().block_timestamp().into()
+        }
+
+        #[ink(message)]
+        pub fn get_balance(&self) -> Balance {
+            self.env().balance()
+        }
+
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: These are not used or needed.

File: contracts/polyx-vesting/lib.rs [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        fn new_init(
+            &mut self,
+            beneficiary_address: AccountId,
+            start_timestamp: u128,
+            duration_seconds: u128,
+        ) {
+            self.beneficiary = beneficiary_address;
+            self.start = start_timestamp;
+            self.duration = duration_seconds;
```
Comment (adamdossa) [inline contracts/polyx-vesting/lib.rs]: Should we validate that `start` and `duration` are sensible? i.e. that `start + duration` doesn't overflow?

File: contracts/polyx-vesting/lib.rs [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+        #[ink(message)]
+        pub fn releasable(&self) -> Balance {
+            self.vested_amount(self.env().block_timestamp().into())
+                .saturating_sub(self.released())
+        }
+
+        /// Release the native token (POLYX) that have already vested.
+        #[ink(message)]
+        pub fn release(&mut self) -> Result<()> {
+            Self::validation(self.start, self.duration)?;
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: This needs to be done in the constructor, not here.

File: contracts/polyx-vesting/lib.rs [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+            duration_seconds: u128,
+        ) {
+            self.beneficiary = beneficiary_address;
+            self.start = start_timestamp;
+            self.duration = duration_seconds;
+        }
+
+        fn validation(start_timestamp: u128, duration_seconds: u128) -> Result<()> {
+            // Ensure start isn't zero
+            if start_timestamp <= 0 || duration_seconds <= 0 {
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: Unsigned integers can't be lower than zero.  Only need to check `== 0`.

Also make sure that the start time is greater then the current block timestamp.

File: contracts/polyx-vesting/lib.rs [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+    /// The contract result type.
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    impl PolyxVesting {
+        /// Constructor
+        #[ink(constructor, payable)]
+        pub fn new(
+            beneficiary_address: AccountId,
+            start_timestamp: u128,
+            duration_seconds: u128,
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: Please use `Timestamp` instead of `u128` for `start`, `duration`.  Timestamps are only u64 values.  No need to use a `u128` for them.

**Review by Neopallium — CHANGES_REQUESTED** [review]

File: contracts/polyx-vesting/lib.rs [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+        }
+
+        fn validation(start_timestamp: u128, duration_seconds: u128) -> Result<()> {
+            // Ensure start isn't zero
+            if start_timestamp <= 0 || duration_seconds <= 0 {
+                return Err(Error::InvalidTimestamp);
+            }
+
+            let year_seconds = 31556926;
+            // Ensure start and duration not over a year
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: I am not sure we should be limiting the duration to just one year.

Comment (Neopallium) [reply contracts/polyx-vesting/lib.rs]: Also I think the block timestamp is in units of Milliseconds, not seconds.

File: contracts/polyx-vesting/lib.rs [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+    impl PolyxVesting {
+        /// Constructor
+        #[ink(constructor, payable)]
+        pub fn new(
+            beneficiary_address: AccountId,
+            start_timestamp: Timestamp,
+            duration_milli_seconds: Timestamp,
+        ) -> Result<Self> {
+            ink_lang::utils::initialize_contract(|contract| {
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: Ink! <=3.4 only supports `Self` as a return type from constructors.  But they have fixed this in version 4.0: https://github.com/paritytech/ink/pull/1446

Polymesh doesn't support Ink! 4.0 until we update the Substrate monthly PR.  So for now just add a comment with a note about this with the link to that Ink! PR and use `unwrap` to return `Self` here.

File: contracts/polyx-vesting/lib.rs [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+            if self.start == 0 || self.duration == 0 {
+                return Err(Error::InvalidTimestamp);
+            }
+
+            let max_milli_seconds = u64::MAX;
+            // Ensure start and duration not over type limit
+            if self.start < ink_env::block_timestamp::<ink_env::DefaultEnvironment>() {
+                return Err(Error::InvalidStartTimestamp);
+            } else if self.duration > max_milli_seconds {
+                return Err(Error::DurationOverflow);
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: The `Timestamp` type is `u64` , so `duration` can't be greater  than `u64::MAX`.  This check will never fail.

Please add a check that `start + duration` doesn't overflow.

File: contracts/polyx-vesting/lib.rs [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+        ) -> Result<()> {
+            self.beneficiary = beneficiary_address;
+            self.start = start_timestamp;
+            self.duration = duration_milli_seconds;
+            self.validation()
+        }
+
+        fn validation(&self) -> Result<()> {
+            // Ensure start isn't zero
+            if self.start == 0 || self.duration == 0 {
```
Comment (Neopallium) [inline contracts/polyx-vesting/lib.rs]: There is already a check to make sure that `start` isn't less than the current block time.  No need to check that it isn't zero.

Also I think it might be ok to allow a zero duration.  If the vesting schedule just wants to use the `start` time to release the total amount.

**Review by Neopallium — CHANGES_REQUESTED** [review]

**Review by Neopallium — APPROVED** [review]

**Review by adamdossa — APPROVED** [review]
